src/lib/list.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1.  Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ * 2.  Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "list.h"
+
+struct list_head {
+    list_item *first, *last;
+    size_t     length;
+};
+
+struct list_item {
+    list_head *head;
+    list_item *prev, *next;
+};
+
+static list_item *
+get_item_ptr(list_item *item)
+{
+    return item ? (item - 1) : NULL;
+}
+
+static void *
+get_data_ptr(list_item *item)
+{
+    return item ? (item + 1) : NULL;
+}
+
+size_t
+list_length(list head)
+{
+    return head ? head->length : 0;
+}
+
+static list_item *
+list_do_insert(list *lst, list_item *where, const void *data, size_t data_size)
+{
+    if (!lst || !data_size) {
+        return NULL;
+    }
+    bool allocated_head = false;
+    if (!*lst) {
+        *lst = calloc(1, sizeof(**lst));
+        allocated_head = true;
+        if (!*lst) {
+            return NULL;
+        }
+    }
+    list head = *lst;
+
+    list_item *item = malloc(sizeof(*item) + data_size);
+    if (!item) {
+        if (allocated_head) {
+            free(*lst);
+            *lst = NULL;
+        }
+        return NULL;
+    }
+    if (data) {
+        memset(item, 0, sizeof(*item));
+        memcpy(get_data_ptr(item), data, data_size);
+    } else {
+        memset(item, 0, sizeof(*item) + data_size);
```

Comment:
I think that `list_do_insert` should return an error when it has `data_size` != 0 and `data` is `NULL`.